pkg/oomstore/apply.go

```diff
@@ -0,0 +1,192 @@
+package oomstore
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/oom-ai/oomstore/internal/database/metadata"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types/apply"
+	"gopkg.in/yaml.v3"
+)
+
+func (s *OomStore) Apply(ctx context.Context, opt apply.ApplyOpt) error {
+	data := make(map[string]interface{})
+	if err := yaml.NewDecoder(opt.R).Decode(data); err != nil {
+		return err
+	}
+
+	var kind string
```

Comment:
## improvement

Shall we add a validation stage before starting to apply? In that stage, we make sure the input config file is valid.

## proposal

To make this cleaner, I wonder if we can parse the input file into a struct

```go
type ApplyGoal struct {
  ...
}
```

This struct contains ALL data we know about the input file. Then, we build a validation method `validate` against the `ApplyGoal` struct, call this method to ensure its validity.

Then, we build a `apply(ApplyGoal)` method that does the apply.

## benefits

More declarative and less procedural, more readable

## cons

Less straightforward

@lianxmfor @jinghancc @wfxr what do you think?

Reply:
sound good! But what do I need to validate? I'm currently thinking of the kind field

Reply:
Hmm maybe as long as the parsing works, then there is little rest for validation.

Reply:
I have refactored according to this improvement plan, but I think the code is a bit ugly...  @YikSanChan @jinghancc @wfxr